### PR TITLE
feat: send mass slack notifications 🧱 

### LINKED ISSUE #188: StackBot Notification Infrastructure 💬
### Description

With the recent Census launch, it was suggested by multiple people that we should utilize the StackBot to notify members about things like survey requests, launches, and more. However, we don't have the infrastructure to send these "one-off" dynamic type of Slack notifications.

This issue will introduce the ability to send mass StackBot notifications (individually to each Slack user).

### Acceptance Criteria

This will be a **backend-only** issue, no need to worry about incorporating into the frontend quite yet (that is to come). Eventually, we'll implement a UI in the Admin Dashboard that accepts a message input (hopefully it'll be a rich text field ??).

Here is what this issue should do:
- Implement the ability to send a StackBot notification with a `message` to all members.

Some things to keep in mind:
- We can only send a StackBot notification to a member if they have a `slackId` linked.
- The notifications themselves should be processed in the background using Bull.

Here is what's **not** in scope:
- Sending a different message to different users (ie: customizing their first name).
- Filtering the members who we send a notification to.

### Additional Context

- You can emit a job using the [`job`](https://github.com/colorstackorg/oyster/blob/f50b54dca44f18644e64e1d44b3b971e27d4c7c0/packages/core/src/infrastructure/bull/use-cases/job.ts#L7) function.
- We already have a [`notification.slack.send`](https://github.com/colorstackorg/oyster/blob/f50b54dca44f18644e64e1d44b3b971e27d4c7c0/packages/core/src/modules/notification/notification.worker.ts#L16) to send the Slack notification.

### PR DESCRIPTION
**Pull request by tomas-salgado**
## Description ✏️

Closes #188 

This PR implements the backend infrastructure for the ability to send mass StackBot notifications (individually to each Slack user). 

Specifically, with this PR we create a `SlackBullJob` named `slack.mass.message`. When triggered, this job will run the function [sendStackBotMassNotification](https://github.com/colorstackorg/oyster/blob/4b83016367a5d033b3542e6a98081ac12377a574/packages/core/src/modules/slack/use-cases/send-stackbot-mass-notification.ts#L5-L19), which uses the job `notification.slack.send` to send a Slack message to all members with a valid `slackId`.

## Type of Change 🐞

- [X] Feature - A non-breaking change which adds functionality.
- [ ] Fix - A non-breaking change which fixes an issue.
- [ ] Refactor - A change that neither fixes a bug nor adds a feature.
- [ ] Documentation - A change only to in-code or markdown documentation.
- [ ] Tests - A change that adds missing unit/integration tests.
- [ ] Chore - A change that is likely none of the above.

## Checklist ✅

- [x] I have done a self-review of my code.
- [ ] I have manually tested my code (if applicable).
- [ ] I have added/updated any relevant documentation (if applicable).

### CODE REVIEW

**Comment by tomas-salgado** [comment]
A few comments:

1. I was trying to understand how the frequency of jobs are set, and from my understanding it seems that for repeatable jobs (which this is not), the frequency is inputted as [pattern](https://github.com/colorstackorg/oyster/blob/fceb18acfe3a40dd0a5d75b719eebcdef2bc0aa6/apps/admin-dashboard/app/routes/_dashboard.bull.%24queue.repeatables.add.tsx#L73). Otherwise for one-off jobs (such as this one), when the job is created on the admin UI, are they just added to the queue to be completed ASAP?  
2. My understanding is that the current state of this PR covers all the backend infra needed. The next steps for the feature would be implementing a frontend that can create a job named `slack.mass.message`.
3. It seems in all my PRs that my old commits (from previous PRs) are still showing. Any idea why this is? 😆

File: packages/core/src/modules/slack/use-cases/send-stackbot-mass-notification.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import { type GetBullJobData } from '@/infrastructure/bull/bull.types';
+import { job } from '@/infrastructure/bull/use-cases/job';
+import { db } from '@/infrastructure/database';
+
+export async function sendStackBotMassNotification(
+  data: GetBullJobData<'slack.mass.message'>
+) {
+  const members = await db.selectFrom('students').select(['slackId']).execute();
```
Comment (ramiAbdou) [inline packages/core/src/modules/slack/use-cases/send-stackbot-mass-notification.ts:8]: I know we're doing the conditional check for `member.slackId` within the for loop, but I think an additional improvement would be adding a `.where('slackId', 'is not', null)` check in the SQL query itself!

**Review by ramiAbdou — CHANGES_REQUESTED** [review]
This PR overall looks really good!

That being said, I don't actually think we need to make this a Bull job (ie: `slack.mass.message`). In our UI, instead of emitting a job from there, I think we can just call the `sendStackBotMassNotification` function directly!

**Comment by ramiAbdou** [comment]
> 1. I was trying to understand how the frequency of jobs are set, and from my understanding it seems that for repeatable jobs (which this is not), the frequency is inputted as [pattern](https://github.com/colorstackorg/oyster/blob/fceb18acfe3a40dd0a5d75b719eebcdef2bc0aa6/apps/admin-dashboard/app/routes/_dashboard.bull.%24queue.repeatables.add.tsx#L73). Otherwise for one-off jobs (such as this one), when the job is created on the admin UI, are they just added to the queue to be completed ASAP?

Don't think we'll need to use Bull anymore for this PR, but to answer your question: when a job is added to the queue, it is indeed completed ASAP! Also, 99.9% of the time, one-off jobs are added from code and not from the Admin Dashboard UI!

> 2. My understanding is that the current state of this PR covers all the backend infra needed. The next steps for the feature would be implementing a frontend that can create a job named `slack.mass.message`.

Correct!

> 3. It seems in all my PRs that my old commits (from previous PRs) are still showing. Any idea why this is? 😆

I would just make sure that your main branch is up to date with the upstream main branch. Rebasing always produces a cleaner more linear history as well:

```sh
git rebase upstream/main
```

File: packages/core/src/modules/slack/use-cases/send-stackbot-mass-notification.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import { job } from '@/infrastructure/bull/use-cases/job';
+import { db } from '@/infrastructure/database';
+
+export async function sendStackBotMassNotification(message: string) {
```
Comment (tomas-salgado) [inline packages/core/src/modules/slack/use-cases/send-stackbot-mass-notification.ts:4]: Changed the parameter from `data` from a Bull job to just a string. Also, I believe this should still support rich text messages so long as the message is passed into this function with the correct formatting.

**Comment by tomas-salgado** [comment]
I removed the implementation of the Bull job and just left the implementation of `sendStackBotMassNotification`. Lmk if anything else is needed!

**Review by ramiAbdou — APPROVED** [review]
Let's ship it!

**Pull request merged**